lib/seattleflu/db/cli/command/etl/longitudinal.py

```diff
@@ -0,0 +1,356 @@
+"""
+Process longitudinal documents into the relational warehouse.
+"""
+import click
+import logging
+from math import ceil
+from datetime import datetime, timezone
+from typing import Any
+from seattleflu.db.session import DatabaseSession
+from seattleflu.db.datatypes import Json
+from . import etl, find_or_create_site, upsert_individual, upsert_encounter
+from . import sample_identifier, find_sample, update_sample
+from . import UnknownSiteError, UnknownRaceError, UnknownEthnicGroupError
+from . import UnknownFluShotResponseError, UnknownInsuranceResponseError
+from .presence_absence import SampleNotFoundError
+
+
+LOG = logging.getLogger(__name__)
+
+
+# This revision number is stored in the processing_log of each longitudinal
+# record when the longitudinal record is successfully processed by this ETL
+# routine. The routine finds new-to-it records to process by looking for
+# longitudinal records lacking this revision number in their log.  If a
+# change to the ETL routine necessitates re-processing all longitudinal records,
+# this revision number should be incremented.
+REVISION = 1
+
+
+@etl.command("longitudinal", help = __doc__)
+
+@click.option("--dry-run", "action",
+    help        = "Only go through the motions of changing the database (default)",
+    flag_value  = "rollback",
+    default     = True)
+
+@click.option("--prompt", "action",
+    help        = "Ask if changes to the database should be saved",
+    flag_value  = "prompt")
+
+@click.option("--commit", "action",
+    help        = "Save changes to the database",
+    flag_value  = "commit")
+
+def etl_longitudinal(*, action: str):
+    LOG.debug(f"Starting the longitudinal ETL routine, revision {REVISION}")
+
+    db = DatabaseSession()
+
+    # Fetch and iterate over longitudinal records that aren't processed
+    #
+    # Rows we fetch are locked for update so that two instances of this
+    # command don't try to process the same longitudinal records.
+    LOG.debug("Fetching unprocessed longitudinal records")
+
+    longitudinal = db.cursor("longitudinal")
+    longitudinal.execute("""
+        select longitudinal_id as id, document
+          from receiving.longitudinal
+         where not processing_log @> %s
+         order by id
+           for update
+        """, (Json([{ "revision": REVISION }]),))
+
+    processed_without_error = None
+
+    try:
+        for record in longitudinal:
+            with db.savepoint(f"longitudinal record {record.id}"):
+                LOG.info(f"Processing longitudinal record {record.id}")
+
+                # Check validity of barcode
+                received_sample_identifier = sample_identifier(db,
```

Comment:
Do we always expect a sample with this longitudinal data? I thought they were doing surveys even if they don't swab the kid?

Reply:
Yes, you're right! I end up with a lot of missing barcodes in the end. In the current process, I'm dropping those records without barcodes. I was operating under the assumption that we only care about encounters with swabs. @tsibley -- is this the case?

Reply:
Ah, no, we do care about encounters without swabs. They'll have encounters with details, which may still be useful.

Reply:
Ok, that's going to require some changes. I'll get started on those.

Reply:
Done with those changes now. I'm not skipping on missing barcode or missing sample, but only on missing encounter date.